posts/frangeris/04.upgradeable-contracts.md

```diff
@@ -0,0 +1,228 @@
+
+Some of the caveats of using the proxy pattern are:
+
+1. Initializers: Contructors are not allowed in upgradeable contracts, instead we need to use the `initialize` function from the `Initializable` contract. This function is called only once when the contract is deployed for the first time. This means that we need to use the `initializer` modifier in the `initialize` function.
+
+2. Extending from upgradeable contracts libraries: Following the same logic, we need to use the `initializer` modifier when extending from upgradeable contracts libraries for libraries that we are extending from.
+
+3. Initial values and fields declarations: We need to be careful when declaring fields in our contracts, because the initial values are not set when the contract is deployed. This means that we need to set the initial values in the `initialize` function.
+
+4. Initialization of implemented contracts: As constructors are disabled when working with upgradeability, while extending from a contract, we call the function that "initialize" instead of calling the parent via `super`.
+
+5. Not `selfdestruct` or `delegatecall` implemented: These functions are not supported in upgradeable contracts to avoid breaking the proxy pattern or opening a door to security issues.
+
+More detailed information can be found in the [OpenZeppelin Upgrades documentation](https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable).
+
+## The implementation
+
+To give the ability of upgradeability to our "Gateway" (in case something goes wrong or need to change), we need to use the proxy pattern. So let's start by creating and deploying a `Gateway.sol` contract:
+
+```solidity
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.18;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+contract Gateway is Initializable, OwnableUpgradeable {
+  uint private fee;
+
+  function initialize(uint _fee) public initializer {
+    fee = _fee;
+    __Ownable_init();
+  }
+
+  function getFee() public view returns (uint) {
+    return fee;
+  }
+}
+```
+
+This contract is very simple, it has a `fee` variable that can be set by the owner of the contract. This is the contract that we are going to be upgrading.
+
+To deploy the contract we need to use the `deployProxy` function from the `@openzeppelin/hardhat-upgrades` plugin. This function will deploy the proxy contract and initialize it with the `initialize` function from the `Gateway` contract.
+
+To keep things simple and visible (in case you cant deploy), we will use [Sepolia testnet](https://sepolia.dev/) blockchain and [Hardhat](https://hardhat.org/) as development environment, so to configure hardhat to use sepolia we need to add the following configuration to the `hardhat.config.js` file:
```

Comment:
```suggestion
To keep things simple and visible (in case you cant deploy), we will use [Sepolia testnet](https://sepolia.dev/) blockchain and [Hardhat](https://hardhat.org/) as development environment, so to configure Hardhat for the use of Sepolia we need to add the following configuration to the `hardhat.config.js` file:
```

Reply:
`for the` is not correct here, sepolia is just a network that hardhat connect to, `hardhat doesn't "makes use" of sepolia`, hardhat "use" sepolia